### PR TITLE
Run swc.minifySync inside a pool of forked child processes

### PR DESCRIPTION
**Pull request by mreishus**
### PR Dependency Note

* This includes the prettier changes as well. If #5 is merged, this PR will be shorter and easier to read.

### Summary

**Fork child processes when doing JS minification. This contains the memory leak.**

* We use a pool of child processes and use the `generic-pool` library to help manage this.
* New files are lin `lib/processes`.

There is also some new optional debugging to help see what's going on, opted in via environment flags.

### Test Plan

- Turn on the new debuggers with this in bash:

```
export DEBUG_MEMORY=1
export DEBUG_POOL=1
export DEBUG_QUIET_REQUEST=1
```

- Run the server
  - `npm start`
- Clone the load tester: https://github.com/mreishus/test-mini-js 
  - In a 2nd terminal window, cd into `files` and run `./serve.sh` in test-mini-js
  - In a 3rd terminal window, run `node main.js` in test-mini-js. This should send lots of requests to the server
- Double check the minifier is working in your browser: Visit `http://127.0.0.1:4747/get?url=http%3A%2F%2F127.0.0.1%3A1234%2Freact.production.min.js&random=123asdf`
- Watch the memory usage. It should not be unbounded.

To compare to the old versions, shut off the server, make this change:
```diff
diff --git a/config.js b/config.js
index c7c9a7c..6458499 100644
--- a/config.js
+++ b/config.js
@@ -14,8 +14,8 @@ module.exports = () => {
                        regex: new RegExp(
                                /^(application\/javascript)|(application\/x-javascript)|(application\/ecmascript)|(text\/javascript)|(text\/ecmascript)/
                        ),
-                       min_func: require( './lib/processes/js-min-parent.js' ), // Forking version
-                       // min_func: require( './lib/js-min.js' ), // Non-forking version
+                       //min_func: require( './lib/processes/js-min-parent.js' ), // Forking version
+                       min_func: require( './lib/js-min.js' ), // Non-forking version
                        type: 'string',
                },
                json: {
```
Re-do the steps, and you should see the memory usage become unbounded.

### CODE REVIEW

**Comment by sgomes** [comment]
Rebased on trunk after merging #5.

File: lib/memory-debugger.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module.exports = function maybeEnableMemoryDebugger() {
+	if ( process.env.DEBUG_MEMORY ) {
+		console.log( 'Memory debugging enabled' );
+		setInterval( () => {
+			const memoryUsage = process.memoryUsage();
+			console.log( '---' );
```
Comment (sgomes) [inline lib/memory-debugger.js]: Suggest using `console.debug()` instead, across this method. It's functionally the same as `console.log()` in node.js, but it's a little bit more semantic.

File: lib/pool-monitor.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module.exports = function maybeEnablePoolMonitor( pool ) {
+	if ( process.env.DEBUG_POOL ) {
+		const printPoolInfo = () => {
+			console.log( 'spareResourceCapacity:', pool.spareResourceCapacity );
```
Comment (sgomes) [inline lib/pool-monitor.js]: Same as other console comment.

File: lib/processes/js-min-child-process-pool.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// This manages a pool of child processes, and is used by js-min-parent.js.
+const { fork } = require( 'child_process' );
```
Comment (sgomes) [inline lib/processes/js-min-child-process-pool.js:2]: This looks like it goes back pretty far in terms of support, so it shouldn't be a problem regardless of what version of node we're running in production (18, according to the docker file).

File: lib/processes/js-min-child-process-pool.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	},
+};
+
+const opts = {
+	min: 4,
+	max: 32,
+	evictionRunIntervalMillis: 10000,
+	softIdleTimeoutMillis: 10000,
+	idleTimeoutMillis: 30000,
+};
```
Comment (sgomes) [inline lib/processes/js-min-child-process-pool.js]: We'll need to come up with a plan to test these values in production, but they seem sane enough at first glance.

File: lib/processes/js-min-parent.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+		child.once( 'message', ( minifiedData ) => {
+			resolve( minifiedData );
+
+			child.count++;
+			if ( child.count >= 50 ) {
+				pool.destroy( child );
+			} else {
+				pool.release( child );
+			}
+		} );
```
Comment (sgomes) [inline lib/processes/js-min-parent.js]: If I'm reading this right, each child is allowed to handle 50 minifications before being destroyed, correct? That sounds sensible enough if so, but we may want to add a comment explaining that that's what's happening here.

**Review by sgomes — APPROVED** [review]
Left some comments, but nothing blocking, the code looks great! 👍 Excellent work, @mreishus! 

I'll put this through some tests and will let you know if anything is off, but feel free to make any changes and commit in the meantime.

**Comment by sgomes** [comment]
Just a note to confirm that this seems to work correctly for me, with memory usage rising before later dropping, as the spawned processes age and are eventually destroyed 👍

File: lib/processes/js-min-parent.js (hunk 5 above)
Comment (mreishus) [reply lib/processes/js-min-parent.js]: Yes, that's correct.. I'm not sure if the number is too high or too low.

File: lib/memory-debugger.js (hunk 1 above)
Comment (mreishus) [reply lib/memory-debugger.js]: Changed in 0aa5350fb38

File: lib/processes/js-min-parent.js (hunk 5 above)
Comment (mreishus) [reply lib/processes/js-min-parent.js]: Moved to constant with comment in 23d97efce4

**Comment by mreishus** [comment]
I think the error handling needs work. Yesterday I was so focused on the happy path that I didn't really pay attention to it.

## Test plan

1. Force both minifiers to fail:

```diff
diff --git a/lib/js-min.js b/lib/js-min.js
index 0a7a2bf..c6d1fa8 100644
--- a/lib/js-min.js
+++ b/lib/js-min.js
@@ -4,7 +4,7 @@
 
 module.exports = ( body ) => {
        const swc = require( '@swc/core' );
-       const result = swc.minifySync( body, {
+       const result = swc.minifySync( body + "asdfasd{asdfasd{", {
                compress: false,
                mangle: true,
        } );
diff --git a/lib/processes/js-min-child.js b/lib/processes/js-min-child.js
index 80e4d44..5a14e94 100644
--- a/lib/processes/js-min-child.js
+++ b/lib/processes/js-min-child.js
@@ -1,7 +1,7 @@
 // This should be forked as a child process.
 process.on( 'message', ( body ) => {
        const swc = require( '@swc/core' );
-       const result = swc.minifySync( body, {
+       const result = swc.minifySync( body + "asdfasd{asdfasd{", {
                compress: false,
                mangle: true,
        } );
```
- Run the server without the debugging environment variables: Start a new shell/window, then `npm start` (I'm not sure if setting them to 0 works currently?)
- Clone the load tester: https://github.com/mreishus/test-mini-js 
  - In a 2nd terminal window, cd into `files` and run `./serve.sh` in test-mini-js
  - Don't run the main load tester though
- Visit `http://127.0.0.1:4747/get?url=http%3A%2F%2F127.0.0.1%3A1234%2Fjquery.js&random=123asd` in your browser, which will now error during the minification process, and compare results on the non-forked and forked versions
- Non-forked:
  -  Browser receives an error immediately `{"status":"error","code":10000,"msg":"Error requesting the original resource"}`
  - Only log messages in the console
- Forked:
  -  Browser never gets a response and that tab seems to hang, waiting.
  - Console shows crash, debugging errors, stack trace
```
            return bindings.minifySync(toBuffer(src), toBuffer(opts !== null && opts !== void 0 ? opts : {}));
                            ^
                               
Error:                                                                                                                       
  × Expected ';', '}' or <eof>
       ╭─[10362:1]
 10362 │ 
 10363 │ return jQuery;
 10364 │ } );                                                 
 10365 │ asdfasd{asdfasd{
       · ───┬────                                                                                                            
       ·    ╰── This is the expression part of an expression statement                            
       ╰────                                                                                                                 
                                                              
                                                              
Caused by:                                                                                                                   
    0: failed to parse input file
    1: Syntax Error
    at Compiler.minifySync (/home/user/minifiers2/node_modules/@swc/core/index.js:107:29)
    at Object.minifySync (/home/user/minifiers2/node_modules/@swc/core/index.js:370:21)
    at process.<anonymous> (/home/user/minifiers2/lib/processes/js-min-child.js:4:21)

```

**Comment by mreishus** [comment]
I improved the error handling by having the child process use try catch, then send special error messages to the parent. I'm still not sure about uncaught exceptions.
```js
		child.removeAllListeners( 'error' );
		child.once( 'error', ( err ) => {
			reject( err );
			pool.destroy( child );
		} );
```
I don't think this is doing anything and can be removed. I would like to find a suitable replacement for uncaught exceptions, although maybe it doesn't matter with the current simple children.

Edit/Update: Actually, the error event is fine, it's just for different types of errors. https://nodejs.org/api/child_process.html#event-error

**Comment by mreishus** [comment]
Added a fix that stops the request from hanging indefinitely if the child process unexpectedly exits. Testable with:
```diff
diff --git a/lib/processes/js-min-child.js b/lib/processes/js-min-child.js
index 47748b9..b625f1d 100644
--- a/lib/processes/js-min-child.js
+++ b/lib/processes/js-min-child.js
@@ -1,6 +1,7 @@
 // This should be forked as a child process.
 process.on( 'message', ( body ) => {
        const swc = require( '@swc/core' );
+       process.exit(0);
        try {
                const result = swc.minifySync( body, {
                        compress: false,
```

**Comment by mreishus** [comment]
Today's round of updates:
 * `MINIFICATIONS_PER_PROCESS` settable by environment
 * Compression can be disabled by `MINIFIERS_DISABLE_COMPRESSION` environment
 * All binary environment flags will turn off if set to 0, false or empty string
 * When compression is not performed, stop sending a blank `Content-Encoding` header
 * If there's an error on a `/file` endpath, the entire server will no longer crash. 
   * This was introduced when I changed the minifier to be async, but the `/file` endpath did not use await in the right spots, so it could not catch any errors.
 * Fixed a bug when compression is turned on, but clients do not accept any compression protocols the server understands (like wget by default). In this case, clients were getting 1 byte responses and strange `Content-Encoding: <2nd character of response>` headers.
 
Added another PR for path verification: #7

File: routes/file.js [hunk 6]
```diff
@@ -133,25 +133,26 @@ module.exports = ( request, reply ) => {
 			// the original version back.
 			do_compress = false;
 		}
+		do_compress = true;
```
Comment (sgomes) [inline routes/file.js]: Did you mean to leave this here? It looks like a stray debugging/testing statement.

Comment (mreishus) [reply routes/file.js]: Oops! Good find, that's exactly what it is. Removed in 6fe3775.

**Pull request merged**